### DISCUSSION TITLE
STM32

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#977] [General]
Naive question: has there been any effort/interest in supporting STM32 MCUs? I'm particularly interested in the STM32WLE5, which combines the cpu and an sx1262 LoRA radio...

**Comment by phoddie** [comment]
There is interest and effort. We'd very much like to support the STM32 family of MCUs. There's a company that has done an initial port of the XS JavaScript to STM32. Unfortunately, that hasn't been published yet, as far as I know.  Porting to an entirely new device is non-trivial, though if you already know the SDK (tools & APIs) for the target device that's a big head-start. Our Raspberry Pi Pico port is a good reference point for new ports.

Regarding LoRa, we have support for an ESP32 LoRa device, the [Heltec LoRa 32](https://github.com/Moddable-OpenSource/moddable/tree/public/build/devices/esp32/targets/heltec_lora_32). That uses the ST127x LoRa part. @cmidgley did the [ST127x LoRa driver](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/st127x/st127x.js), starting from an Arduino implementation and reworking it into an ECMA-419 style peripheral. It works very nicely. If you want to experiment with LoRa in the Moddable SDK (and onward to Node-RED!) that's an easy starting point.

**Reply by tve** [reply]
If you can find out more about the state of the STM32 port you mention I would appreciate!

**Reply by phoddie** [reply]
Sure, I will ask about that.

Out of curiosity, how much experience do you have with native development on STM32? Do you happen to know if it supports a GCC toolchain?

**Reply by tve** [reply]
I'm not an STM32 whizz but I have written complete bare-bones C/C++ code for stm32l0 series. I have much less experience with the standard software (every time I look at the reams and reams of include files I vomit and open the ref man and code to the hw bits). From a practical standpoint I was thinking of using libopencm3.
GCC definitely supports STM32 (hmm, it's ARM, I don't think there's something special for stm32 in gcc).

**Reply by phoddie** [reply]
Yes, the SDKs from the silicon vendors can be overwhelming to get started with. We've been through that plenty of times. Fortunately, we've always gotten through that.

A GCC toolchain definitely simplifies things. The first big step in a fresh silicon port is building, and with GCC there are good models to follow in the Moddable SDK already.

**Comment by louisvangeldrop** [comment]
Maybe the byte alliance wamr project is an option. Moddable already supports a mcconfig to wasm.
